### ISSUE TITLE
FR: Sleep screen message "shrink to fit" 

### ISSUE THREAD

**Issue by Commodore64user** [koreader/koreader#12109]
**Does your feature request involve difficulty completing a task? Please describe.**

A picture is worth a thousand words...

<img src="https://github.com/koreader/koreader/assets/97603719/fa598ee9-dc49-4f9a-a9ce-8a11f1311e9e" width=50% height=50%>

**Describe the solution you'd like**
An automatic option or setting to reduce text size so that the sleep screen message fits on a single line.

**Comment by christian-ho** [comment]
Or it could use the complete width, like it does if placed at the top or at the bottom.

**Comment by Commodore64user** [comment]
> Or it could use the complete width, like it does if placed at the top or at the bottom.

I don’t think it uses the whole width though, just 90% of it.

**Comment by Frenzie** [comment]
I believe that was the point. ;-)

**Comment by Commodore64user** [comment]
Yes i get that, the thing about the middle one is that it uses the info box so that would need to change

**Comment by Commodore64user** [comment]
https://github.com/koreader/koreader/blob/ca569c1355d14ad023eb1bcad7b995ae027e423d/frontend/ui/screensaver.lua#L642-L647

**Comment by Commodore64user** [comment]
https://github.com/koreader/koreader/blob/5d2782ddfefe63ca2007c4e9019bc74c0ec6efcb/frontend/ui/widget/menu.lua#L154-L167

https://github.com/koreader/koreader/blob/5d2782ddfefe63ca2007c4e9019bc74c0ec6efcb/frontend/ui/widget/textwidget.lua#L67-L88

**Comment by Commodore64user** [comment]
Not quite shrink to fit, but I think I found a solution although it sacrifices some padding:

<p align="center">
<img src="https://github.com/user-attachments/assets/9c6a1b76-231e-4095-a845-36744b13d165" width=40% height=40%> 
<img src="https://github.com/user-attachments/assets/e436ce62-37ac-4420-8816-7000699e98a6" width=40% height=40%> 
</p>

```lua
elseif self.limit_to_one_line then
        text_widget = TextWidget:new{
            text = self.text,
            face = self.face,
            max_width = 0.8 * Screen:getWidth(),
            lang = self.lang,
            para_direction_rtl = self.para_direction_rtl,
            auto_para_direction = self.auto_para_direction,
        }
```

**Comment by hius07** [comment]
If unneeded, the InfoMessage i icon can be removed by `show_icon = false`.

**Comment by Commodore64user** [comment]
@poire-z  might want to pass judgment here...

**Comment by poire-z** [comment]
Pass judgement = decide between 2 things ? Which ?
Or suggest some ideas to solve the issue ?
No real idea.
Looking at the bit of code involved, I think:
https://github.com/koreader/koreader/blob/42d3f25034778c121dc458b3d5246896fedc5db2/frontend/ui/screensaver.lua#L641-L675
I would just remove the special case of "middle" using InfoMessage (so, don't try to fix this in InfoMessage as you'll be solving only this case and not the others) - and use a TextBoxWidget like in the other branch.
Then you can check the height or number of lines of the TextBoxWidget, and loop by reducing the font face until you reach 1 line. We must have that kind of loop here and there, like in InfoMessage or plugins/coverbrowser.koplugin/listmenu.lua.

**Comment by Commodore64user** [comment]
> I would just remove the special case of "middle" using InfoMessage (so, don't try to fix this in InfoMessage as you'll be solving only this case and not the others) - and use a TextBoxWidget like in the other branch.
Then you can check the height or number of lines of the TextBoxWidget, and loop by reducing the font face until you reach 1 line. We must have that kind of loop here and there, like in InfoMessage or plugins/coverbrowser.koplugin/listmenu.lua.

But that is the only case that needs fixing... Both top and bottom can fit a large number of characters into a single line which I am happy to deem fit for purpose.

The thing about using `TextWidget` https://github.com/koreader/koreader/issues/12109#issuecomment-2430448135 is that it does not keep a consistent size box depending on text length, therefore the box varies in size as seen in those two samples. And it also kills all padding.

**Comment by poire-z** [comment]
> But that is the only case that needs fixing... Both top and bottom can fit a large number of characters into a single line which I am happy to deem fit for purpose.

So, kill that case that needs fixing :) and make it instead work like the other cases.
ie. try it, may work:
```diff
+local CenterContainer = require("ui/widget/container/centercontainer")
[...]
-  if message_pos == "middle" then 
+  if false and message_pos == "middle" then 
[...]
     if message_pos == "bottom" then 
         container = BottomContainer 
+    elseif message_pos == "middle" then 
+        container = CenterContainer 
     else 
         container = TopContainer 
     end 
```

**Comment by Commodore64user** [comment]
but I quite like the info box, I really don't like that banner thing (which is why I use neither top nor bottom messages).

**Comment by Commodore64user** [comment]
so this is what I have come up with https://github.com/koreader/koreader/compare/master...Commodore64user:KOReader_fork:one-line-msg?expand=1 thoughts... it forces the single line, still uses the info message box.

**Comment by poire-z** [comment]
Not fond of that: you're travesting InfoMessage to be/do what it is not.
What do you like about the centered InfoMessage? The icon, the border, the rounded corners, the padding ? You could hack message_widget with what you like about it in screensaver.lua - if you don't need all the InfoMessage features (dismissable, movable, multilines auto sizing or font reducing).

**Comment by Commodore64user** [comment]
sorry, I had forgotten about this. well it is probably that I dislike it less than the other two.  Is there a wa in which we could simply force it to not wrap around text even if it needs to sacrifice some padding or something
